src/url-loader/url-resolver.ts

```diff
@@ -28,7 +31,32 @@ export abstract class UrlResolver {
    *
    * Returns `undefined` if the given url cannot be resolved.
    */
-  abstract resolve(url: PackageRelativeUrl): ResolvedUrl|undefined;
+  abstract resolve(
+      url: FileRelativeUrl, baseUrl?: ResolvedUrl,
+      _scannedImport?: ScannedImport): ResolvedUrl|undefined;
+
+  abstract relative(to: ResolvedUrl): FileRelativeUrl;
+  abstract relative(from: ResolvedUrl, to: ResolvedUrl): FileRelativeUrl;
+  abstract relative(from: ResolvedUrl, to: ResolvedUrl, kind: string):
+      FileRelativeUrl;
+  abstract relative(from: ResolvedUrl, to?: ResolvedUrl, kind?: string):
+      FileRelativeUrl;
+
+  protected simpleUrlRelative(from: ResolvedUrl, to: ResolvedUrl):
+      FileRelativeUrl {
+    if (!from.endsWith('/')) {
+      from = this.brandAsResolved(posix.dirname(from));
+    }
+    let result = posix.relative(from, to);
+    if (to.endsWith('/')) {
+      result += '/';
+    }
+    return this.brandAsRelative(result);
+  }
+
+  protected brandAsRelative(url: string): FileRelativeUrl {
```

Comment:
nit: maybe just `asRelative`? "brand" is mostly the name of the technique

Reply:
I wanted to make it clear that this does no logic. It's just taking a string and stamping it as the other type.

(I prefer this over a cast because casts are too loose)